### PR TITLE
Use separate Bc/Gmp math classes.

### PR DESCRIPTION
**Pull request by jwpage**
I noticed that the current `Hashids\Math` class is making a function_exists check with each function call, which was quite slow. I've replaced it with a `Hashids\Math\MathFactory` that checks if the relevant extension (gmp/bcmath) is loaded instead, which returns a `Hashids\Math\MathInterace` class.

I've tried to retain backwards compatibility by using the `MathFactory` inside the (deprecated, no longer used) `Hashids\Math` class as well.

Here are a couple of blackfire.io profiles for [encoding 50,000 hashes](https://gist.github.com/jwpage/ad9e1461b2f78b0c64247a6292f490f6).

* Before: [8.61s](https://blackfire.io/profiles/5ba004bf-d8b5-4966-a927-6a09314cdbe2/graph)
* After: [6.06s](https://blackfire.io/profiles/46c22cb8-39fc-4183-aebe-f879795515ee/graph)

----

🚨 A couple of things that need fixing/addressing:

* `Hashids\Math\Gmp` and `Hashids\Math\Bc` can both be tested directly in the standard testsuite (`tests/Math/MathInterfaceTest.php`), so the disable_functions  options in `.travis.yml` have been removed.
* I couldn't figure out a way to disable extensions at runtime (either command-line or in PHP), so `Hashids\Math\MathFactory` is marked as `@codeCoverageIgnore`. 👎 (Considered adding a bitmask parameter to the constructor to force disabling GMP and/or BC, but is it worth it?)
* `tests/Hashids/RuntimeExceptionTest.php` also can't be run properly, as there's no way to force disabling extensions at runtime. Could this test be removed instead? 🤷‍♂️

### CODE REVIEW

**Comment by ryan-senn** [comment]
Probably doesn't really matter, but I'd make the methods non-static at that point. Feels odd to have a factory and then not use the created instance.

**Comment by vinkla** [comment]
Thanks for the pull request @jwpage! It is awesome that you're taking the time to optimise this library. 

There are a couple of thing we need to adjust before showing this to @ivanakimov:

- Add docblocks to all methods
- Add you name and email to the classes you've adjusted
- Fix the test suite to return green status

We should also note that this change will require a new major release since this is a breaking change.

> Probably doesn't really matter, but I'd make the methods non-static at that point. Feels odd to have a factory and then not use the created instance.

I agree, we should make the methods non-static.

> I couldn't figure out a way to disable extensions at runtime (either command-line or in PHP), so Hashids\Math\MathFactory is marked as @codeCoverageIgnore. 👎 (Considered adding a bitmask parameter to the constructor to force disabling GMP and/or BC, but is it worth it?)

Hmm, not sure. In what case would we want to enforce this? Lets skip that for now.

> `tests/Hashids/RuntimeExceptionTest.php` also can't be run properly, as there's no way to force disabling extensions at runtime. Could this test be removed instead? 🤷‍♂️

Lets drop the test entirely for now.

**Comment by jwpage** [comment]
Thanks for the feedback, @vinkla. Hopefully have ticked off all the the issues.

- [x] Add docblocks to all methods
- [x] Add your name and email to the classes you've adjusted
- [x] Fix the test suite to return green status (see also: Remove RuntimeExceptionTest)
- [x] Make methods non-static
- [x] Leave @codeCoverageIgnore in place on MathFactory
- [x] Remove RuntimeExceptionTest

If this is going to be a new major version, I'll go ahead and remove the `Hashids\Math` class as it's no longer used. I'd also maybe recommend adding `@api/@internal` docblock tags to indicate what Structural Elements are supported for future backwards compatibility (only `Hashids::__construct`, `Hashids::encode`, `Hashids::decode`?)

**Comment by vinkla** [comment]
> I'd also maybe recommend adding @api/@internal docblock tags to indicate what Structural Elements are supported for future backwards compatibility (only Hashids::__construct, Hashids::encode, Hashids::decode?)

Yes, sounds like a good idea.

File: src/Hashids.php [hunk 1]
```diff
@@ -75,6 +79,13 @@ class Hashids implements HashidsInterface
      */
     protected $salt;
 
+    /**
+     * The math class.
+     *
+     * @var MathInterface
```
Comment (vinkla) [inline src/Hashids.php]: Please use the full namespaced path.

File: src/Hashids.php [hunk 2]
```diff
@@ -91,6 +102,7 @@ public function __construct($salt = '', $minHashLength = 0, $alphabet = 'abcdefg
         $this->salt = $salt;
         $this->minHashLength = $minHashLength;
         $this->alphabet = implode('', array_unique(str_split($alphabet)));
+        $this->math = MathFactory::create();
```
Comment (vinkla) [inline src/Hashids.php]: Does the `::create()` method have to be static? Maybe we should convert it to just create a new `MathFactory` instance?

Comment (jwpage) [reply src/Hashids.php]: The only reason `MathFactory` existed was so I could re-use the logic in `Hashids\Math`, but now that it's being removed I can just move the `::create` logic directly into `Hashids\Hashids`.

**Comment by jwpage** [comment]
> Sounds like a good idea.

I might leave @api/@internal tagging for a different 3.0-related PR, as that's not directly related to this PR.

**Comment by ivanakimov** [comment]
This is great :smile:

@jwpage, I agree with @vinkla -- thanks for taking the time to optimize the code!
And @vinkla, thanks for such quick responses and taking the lead :muscle: Feel free to release whenever you're ready!

**Comment by vinkla** [comment]
Thanks for the pull request @jwpage! I'll bump the branch alias to 3.0 and then we can test the new version a couple of days before releasing it. Let me know if you've any questions!

**Pull request merged**